roachprod: add start-tenant command

RaduBerinde opened the pull request:
This commit adds a new `start-tenant` command, which starts a set of
SQL instances for a tenant, against a given host cluster.

For example:
```
roachprod create local-host -n 3
roachprod start local-host
roachprod create local-tenant -n 2
roachprod start-tenant local-tenant --host-cluster local-host
roachprod sql local-tenant
```

Note that non-overlapping subsets of nodes of the same cluster can be
used as well (but it can be more confusing):
```
roachprod create local
roachprod start local:1-2
roachprod start-tenant local:3-4 --host-cluster local:1-2
roachprod sql local:3
```

Multiple tenants can be started by passing different `--tenant-id`
(which defaults to 2).

In future work, we will add the capability of starting a tenant proxy,
as well as higher-level commands for creating and starting an entire
host+tenants deployment.

Note that only insecure clusters are supported for now.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73266)
<!-- Reviewable:end -->

rail reviewed (approved):
Nice! There are some nits, but in overall it looks amazing!

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-:-Mple9poEaZZV1e2I3nE:b93vnt3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @healthy-pod, @RaduBerinde, @stevendanna, and @tbg)

___
*[pkg/cmd/roachprod/flags.go, line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-Mpld0u0Cg-wUQyt7ds2:-Mpld0u0Cg-wUQyt7ds3:bjje4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/cmd/roachprod/flags.go#L190)):*
> ```Go
> 	startTenantCmd.Flags().StringVarP(&hostCluster,
> 		"host-cluster", "H", "", "host cluster")
> 	startTenantCmd.MarkFlagRequired("host-cluster")
> ```

The linter complains and wants us to check the error here.
___
*[pkg/cmd/roachprod/main.go, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-MplaMhdBSO-rCYvlEb0:-MplaMhdBSO-rCYvlEb1:b-cqyn0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/cmd/roachprod/main.go#L448)):*
> ```Go
> 
> As a debugging aid, the --sequential flag starts the nodes sequentially so node
> IDs match hostnames. Otherwise nodes are started are parallel.
> ```

Nit: "in parallel".
___
*[pkg/roachprod/multitenant.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-MplbXtiARSTUG4gOfiE:-MplbXtiARSTUG4gOfiF:bt1zfqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/roachprod/multitenant.go#L21)):*
> ```Go
> )
> 
> func StartTenant(
> ```

The linter complains about missing comments ;)
___
*[pkg/roachprod/install/cockroach.go, line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-Mplbisb3vg12Wu0ExHu:-Mplbisb3vg12Wu0ExHv:b-qhocrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/roachprod/install/cockroach.go#L305)):*
> ```Go
> }
> 
> // SQL runs a `cockroach sql` command. It is assumed that the args include the
> ```

A nit: s/SQL/RunSQL/
___
*[pkg/roachprod/install/cockroach.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-MplcCAB-a_SwjaaE0Sl:-MplcCAB-a_SwjaaE0Sm:b-os9ol6) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/roachprod/install/cockroach.go#L473)):*
> ```Go
> 
> 	case StartTenantProxy:
> 		args = []string{"mt", "start-proxy"}
> ```

The linter complains about this not being used.
___
*[pkg/roachprod/install/cockroach.go, line 558 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-Mple-Ki3d4oT9q4DpIW:-Mple-Kj2Kt7OY2opQ2X:b-u2vowg) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/roachprod/install/cockroach.go#L558)):*
> ```Go
> func (c *SyncedCluster) generateStartFlagsKV(
> 	node Node, startOpts StartOpts, vers *version.Version,
> ) []string {
> ```

Nice to see these factored out! \o/


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
roachprod: add start-tenant command

This commit adds a new `start-tenant` command, which starts a set of
SQL instances for a tenant, against a given host cluster.

For example:
```
roachprod create local-host -n 3
roachprod start local-host
roachprod create local-tenant -n 2
roachprod start-tenant local-tenant --host-cluster local-host
roachprod sql local-tenant
```

Note that non-overlapping subsets of nodes of the same cluster can be
used as well (but it can be more confusing):
```
roachprod create local
roachprod start local:1-2
roachprod start-tenant local:3-4 --host-cluster local:1-2
roachprod sql local:3
```

Multiple tenants can be started by passing different `--tenant-id`
(which defaults to 2).

In future work, we will add the capability of starting a tenant proxy,
as well as higher-level commands for creating and starting an entire
host+tenants deployment.

Note that only insecure clusters are supported for now.

Release note: None

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-:-MpnBD0vD5b6J5sMTYka:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @healthy-pod, @rail, @stevendanna, and @tbg)

___
*[pkg/cmd/roachprod/flags.go, line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-Mpld0u0Cg-wUQyt7ds2:-Mpn9o2uC912-W1OO4BT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/cmd/roachprod/flags.go#L190)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

The linter complains and wants us to check the error here.
</blockquote></details>

Done.
___
*[pkg/roachprod/multitenant.go, line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-MplbXtiARSTUG4gOfiE:-MpnA_084ftmVKuKc8Pd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/roachprod/multitenant.go#L21)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

The linter complains about missing comments ;)
</blockquote></details>

Done.
___
*[pkg/roachprod/install/cockroach.go, line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-Mplbisb3vg12Wu0ExHu:-MpnAf5XA8Q-tvadSMrZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/roachprod/install/cockroach.go#L305)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

A nit: s/SQL/RunSQL/
</blockquote></details>

Done.
___
*[pkg/roachprod/install/cockroach.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73266#-MplcCAB-a_SwjaaE0Sl:-MpnAjieCq7qEJ1-oKlb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/546a33802b62ec6008996f1b8e4c1ca08d6820f1/pkg/roachprod/install/cockroach.go#L473)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

The linter complains about this not being used.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
